### PR TITLE
Feature/foa fast iteration

### PR DESCRIPTION
**Pull request by joaquintides**
Changed iterator increment code for "regular layout" group implementations (basically, those for SSE2 and Neon). Iteration speed improves dramatically, look for "Running erasure" in:

https://github.com/boostorg/boost_unordered_benchmarks/commit/46129697b05fab67c42cfa2b918a6e40af26d6e6

### CODE REVIEW

**Comment by pdimov** [comment]
Looks good to me, although I'm not sure we should be merging to develop now. Maybe wait for the beta to go out in case there are critical fixes we need to apply.

**Comment by joaquintides** [comment]
The release notes added in the PR target 1.83, I just submitted it because it's ready but we can wait of course.

File: include/boost/unordered/detail/foa.hpp [hunk 1]
```diff
@@ -861,32 +886,64 @@ class table_iterator
...
   {
-    std::size_t off=reinterpret_cast<uintptr_t>(pc)%sizeof(Group);
-    pc-=off;
-    return off;
+    increment(std::integral_constant<bool,regular_layout>{});
   }
 
-  inline void increment()noexcept
+  inline void increment(std::true_type /* regular layout */)noexcept
+  {
```
Comment (cmazakas) [inline include/boost/unordered/detail/foa.hpp:899]: Should we add a `BOOST_ASSERT(p);` here?

**Comment by joaquintides** [comment]
Latest commit (1c5640cfbe84213b9d2a7eff68e2db744cfa3817) didn't affect performance:
https://github.com/boostorg/boost_unordered_benchmarks/commit/90e178376ffefe73d3e6fc7fa24ef70a26d429a6

**Comment by cmazakas** [comment]
> Latest commit ([1c5640c](https://github.com/boostorg/unordered/commit/1c5640cfbe84213b9d2a7eff68e2db744cfa3817)) didn't affect performance: [boostorg/boost_unordered_benchmarks@90e1783](https://github.com/boostorg/boost_unordered_benchmarks/commit/90e178376ffefe73d3e6fc7fa24ef70a26d429a6)

Well, my logic was more about the code self-documenting the assumptions.

It wasn't clear to me why we needed the null pointer check before calling `++it`.

**Comment by pdimov** [comment]
We can merge it to develop, I suppose, just not to master.

**Pull request merged**